File: leet_easy_c++/leet_303/leet_303a.cpp
class NumArray {
public:
    NumArray(vector<int> nums) {
        len = nums.size();
        num_ary = new int[len*len];
        for(int i=0; i<len; i++) num_ary[i*len + i] = nums[i];
        for(int i=0; i<len-1; i++)
            for(int j=i+1; j<len; j++)
                num_ary[i*len + j] = nums[j] + num_ary[i*len + j-1];
    }
    
    int sumRange(int i, int j) {
        if(i > j){
            int tmp = i;
            i = j;
            j = tmp;
        }
        return num_ary[i*len + j];
    }
private:
    int* num_ary;
    int len;
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(i,j);
 */

File: leet_easy_java/leet_303/leet_303a.java
class NumArray {
    private int[] num_ary;
    
    public NumArray(int[] nums) {
        num_ary = new int[nums.length + 1];
        num_ary[0] = 0;
        for(int i=0; i<nums.length; i++) num_ary[i+1] = num_ary[i] + nums[i];
    }
    
    public int sumRange(int i, int j) {
        return num_ary[j+1]-num_ary[i];
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(i,j);
 */